client/src/elm/User/Model.elm

```diff
@@ -1,14 +1,22 @@
-module User.Model exposing (..)
+module User.Model
+    exposing
+        ( CurrentUser(..)
+        , User
+        , UserTuple
+        )
 
--- @todo: Will be used with EveryDictList
--- type UserId
---     = UserId Int
+import Backend.Entities exposing (UserId)
 
 
-type alias UserId =
-    Int
+type CurrentUser
```

Comment:
@rgrempel if it was a fully decoupled app I would make it:

```elm
type CurrentUser 
    = Anonymous (WebData AccessToken) (WebData ()) 
     | Authenticated  AccessToken UserTuple
```

So while we are fetching the access token or the user based on the access token, they are still kept as Anon. What do you think?

Reply:
I think it makes a lot of sense to keep the access token and the user together in a single data structure, because logically we shouldn't have one without the other -- if we've got an access token, logically we should have the user as well. I did something like that on an old Unio PR that got side-tracked a while ago:

https://github.com/Gizra/unio/pull/1551

In particular, look at this file ... I was even putting the `backendUrl` into the structure (as well as access token and user), and calling it `Credentials`.

https://github.com/Gizra/unio/blob/139b0939f9d61115ca034c63d65f0e543ffb1123/client/src/elm/User/Model.elm

Once Unio heats up again, I should return to that PR sometime and finish it.

Anyway, that does mean that we'd consider the user "logged out" until we get both the access token and the user information, but I think that's a feature rather than a bug -- it simplifies everything else we do (e.g. the arguments to `update`), because we can assume that we either have a full set of credentials, or we don't -- we don't have to deal with intermediate cases. (Except in the fetch itself, but we can feed intermediate values through via `Task.andThen` or something like that, so we end up with full values once all the fetches have completed).

Ideally, I'd try to avoid the `Anonymous (WebData AccessToken) (WebData ())` and just make that part `Anonymous`. I think I can see why you're doing it (you need to track the requests for access token and user while in flight), but I think it could probably be avoided via judicious use of `Task.andThen` to consolidate the multiple requests and feed them back into the program as a unified response.

Reply:
> if we've got an access token, logically we should have the user as well

Not always. Say you go back to your computer after a week. The access token is there in the localStorage, but it might no longer be valid.

> and calling it Credentials

I think having the `backendUrl` in the credentials is wrong -- it's part of the `Config`.

> Anyway, that does mean that we'd consider the user "logged out" until we get both the access token and the user information

But that's indeed the case. Until you are not fully authorized we can't know for sure you are a user. Once we have that info we bump you to be `Authenticated`

> but I think it could probably be avoided via judicious use of Task.andThen

I'd be interested in seeing this :)

Reply:
>> if we've got an access token, logically we should have the user as well
> Not always. Say you go back to your computer after a week. The access token is there in the localStorage, but it might no longer be valid.

Yes, but in that case we shouldn't consider the user valid either. So, they still live and die together -- which is what makes it logical to have them in the same record.

>>and calling it Credentials
>
> I think having the backendUrl in the credentials is wrong -- it's part of the Config.

Ah, indeed it still needs to be in `Config`, since we have to get it from somewhere in order to know how to do login in the first place. However, once we've logged it, it may make sense to pass it around as a function parameter in a `Credentials` structure, along with the access token and user information. In principle, the access_token and user have a relationship to the backendUrl -- it is, after all, an access token for that very backendUrl, so there is a sense in which they logically belong together. And, in practice, it reduces the number of parameters to our functions, so helps reduce clutter. 

Of course, there are a few cases where we really do need to provide the backendUrl by itself, which is fine. And, it helps by identifying -- in the function signature itself -- which functions require Credentials, and which require only a backendUrl.